iroha_config/iroha_config_derive/src/lib.rs

```diff
@@ -0,0 +1,498 @@
+#![allow(clippy::module_name_repetitions, missing_docs)]
+
+//! Module with Configurable derive macro
+
+use proc_macro::TokenStream;
+use proc_macro2::Span;
+use proc_macro_error::abort;
+use quote::quote;
+use syn::{
+    parse::{Parse, ParseStream},
+    Data, DataStruct, DeriveInput, Fields, GenericArgument, Ident, Lit, LitStr, Meta,
+    PathArguments, Token, Type, TypePath,
+};
+
+struct EnvPrefix {
+    _ident: Ident,
+    _eq: Token![=],
+    prefix: LitStr,
+}
+
+mod attrs {
+    pub const ENV_PREFIX: &str = "env_prefix";
+    pub const SERDE_AS_STR: &str = "serde_as_str";
+    pub const INNER: &str = "inner";
+}
+
+fn get_type_argument<'a, 'b>(s: &'a str, ty: &'b Type) -> Option<&'b GenericArgument> {
+    let ty = if let Type::Path(ty) = ty {
+        ty
+    } else {
+        return None;
+    };
+    let segments = &ty.path.segments;
+    if segments.len() != 1 || segments[0].ident != s {
+        return None;
+    }
+
+    if let PathArguments::AngleBracketed(ref bracketed_arguments) = segments[0].arguments {
+        if bracketed_arguments.args.len() == 1 {
+            return Some(&bracketed_arguments.args[0]);
+        }
+    }
+    None
+}
+
+fn is_arc_rwlock(ty: &Type) -> bool {
+    let ty = get_type_argument("Arc", ty)
+        .and_then(|ty| {
+            if let GenericArgument::Type(ty) = ty {
+                Some(ty)
+            } else {
+                None
+            }
+        })
+        .unwrap_or(ty);
+    get_type_argument("RwLock", ty).is_some()
+}
+
+// TODO: make it const generic type once it will be stabilized
+fn parse_const_ident(input: ParseStream, ident: &'static str) -> syn::Result<Ident> {
+    let parse_ident: Ident = input.parse()?;
+    if parse_ident == ident {
+        Ok(parse_ident)
+    } else {
+        Err(syn::Error::new_spanned(parse_ident, "Unknown ident"))
+    }
+}
+
+impl Parse for EnvPrefix {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        Ok(Self {
+            _ident: parse_const_ident(input, attrs::ENV_PREFIX)?,
+            _eq: input.parse()?,
+            prefix: input.parse()?,
+        })
+    }
+}
+
+struct Inner {
+    _ident: Ident,
+}
+
+impl Parse for Inner {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        Ok(Self {
+            _ident: parse_const_ident(input, attrs::INNER)?,
+        })
+    }
+}
+
+struct SerdeAsStr {
+    _ident: Ident,
+}
+
+impl Parse for SerdeAsStr {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        Ok(Self {
+            _ident: parse_const_ident(input, attrs::SERDE_AS_STR)?,
+        })
+    }
+}
+
+#[proc_macro_derive(Configurable, attributes(config))]
+pub fn configurable_derive(input: TokenStream) -> TokenStream {
+    let ast = syn::parse(input).expect("Failed to parse input Token Stream.");
+    impl_configurable(&ast)
+}
+
+fn impl_load_env(
+    ty: &Ident,
```

Comment:
I would suggest to use `type_ident` full name as per our codestyle

Reply:
Should we? It is everywhere in syn public api. For example:

https://docs.rs/syn/1.0.68/syn/struct.Field.html#structfield.ty

It is quite of idiom to me

Reply:
Ok, I guess